File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9R.scala
package TextbookTrack.Chapter09

object BA9R {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  final case class Node(id: Int, parentId: Int, depth: Int)
  final case class Edge(nodeTo: Int, startIx: Int, length: Int)

  class SuffixTree(text: String, suffixArray: List[Int], lcpArray: List[Int]) {
    private val size: Int = text.length
    private val nodeIdGenerator: Iterator[Int] = Iterator.from(0)
    private val adjacencyList: Map[Int, List[Edge]] = buildSuffixTree()

    def edges(): Iterator[String] =
      adjacencyList
        .valuesIterator
        .flatMap(_.map{ case Edge(_, startIx, length) => text.slice(startIx, startIx + length) })

    private def buildSuffixTree(): Map[Int, List[Edge]] = {
      val adjacencyList: MutableMap[Int, List[Edge]] = MutableMap()
      val nodes: MutableMap[Int, Node] = MutableMap()

      @tailrec
      def findStartingNodeForNextEdge(node: Node, lcp: Int): Node = {
        if (node.depth > lcp) findStartingNodeForNextEdge(nodes(node.parentId), lcp)
        else node
      }

      @tailrec
      def loop(xs: List[(Int, Int)], currentNode: Node): Map[Int, List[Edge]] = xs match {
        case Nil => adjacencyList.toMap
        case (suffixStart, lcp) :: xss =>
          val node: Node = findStartingNodeForNextEdge(currentNode, lcp)
          val substringStart: Int = suffixStart + node.depth
          if (node.depth == lcp) loop(xss, addLeaf(adjacencyList, node, nodes, substringStart))
          else {
            val edge: Edge = identifyEdge(adjacencyList, node, substringStart)
            val offset: Int = lcp - node.depth
            val middleNode: Node = splitEdge(node, edge, offset, nodes, adjacencyList)
            loop(xss, addLeaf(adjacencyList, middleNode, nodes, substringStart + offset))
          }
      }

      val rootId: Int = nodeIdGenerator.next()
      val root: Node = Node(rootId, -1, 0)
      nodes(rootId) = root
      loop(suffixArray.zip(lcpArray), root)
    }

    private def addLeaf(adjacencyList: MutableMap[Int, List[Edge]],
                        node: Node,
                        nodes: MutableMap[Int, Node],
                        substringStart: Int): Node = {
      val leafId: Int = nodeIdGenerator.next()
      val labelLength: Int = size - substringStart
      val leaf: Node = Node(id = leafId, parentId = node.id, depth = node.depth + labelLength)
      nodes(leafId) = leaf
      adjacencyList.update(node.id, Edge(leafId, substringStart, labelLength) :: adjacencyList.getOrElse(node.id, Nil))
      leaf
    }

    private def identifyEdge(adjacencyList: MutableMap[Int, List[Edge]], node: Node, substringStart: Int): Edge = {
      val char: Char = text(substringStart)
      adjacencyList(node.id).find{ case Edge(_, startIx, _) => text(startIx) == char } match {
        case Some(edge) => edge
        case None => throw new Exception("Edge not found.")
      }
    }

    private def splitEdge(node: Node,
                          edge: Edge,
                          offset: Int,
                          nodes: MutableMap[Int, Node],
                          adjacencyList: MutableMap[Int, List[Edge]]): Node = {
      val endNode: Node = nodes(edge.nodeTo)
      val middleNodeId: Int = nodeIdGenerator.next()
      val middleNode: Node = Node(id = middleNodeId, parentId = node.id, depth = node.depth + offset)
      nodes(middleNodeId) = middleNode
      nodes(endNode.id) = endNode.copy(parentId = middleNodeId)
      val firstEdge: Edge = Edge(nodeTo = middleNodeId, startIx = edge.startIx, length = offset)
      val secondEdge: Edge = Edge(nodeTo = edge.nodeTo, startIx = edge.startIx + offset, length = edge.length - offset)
      adjacencyList.update(middleNodeId, secondEdge :: adjacencyList.getOrElse(middleNodeId, Nil))
      adjacencyList.update(node.id, firstEdge :: adjacencyList.getOrElse(node.id, Nil))
      adjacencyList.update(node.id, adjacencyList(node.id).filterNot(_ == edge))
      middleNode
    }
  }

  private def convertToIntlist(line: String, sep: String = ", "): List[Int] =
    line.split(sep).map(_.toInt).toList

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val suffixArray: List[Int] = convertToIntlist(reader.next())
    val lcpArray: List[Int] = convertToIntlist(reader.next())
    val suffixTree = new SuffixTree(text, suffixArray, lcpArray)
    suffixTree.edges().foreach(println)
  }
}

File: Python/textbook_track/chapter09/ba9r.py
# Construct a Suffix Tree from a Suffix Array
import itertools as it
import sys

from collections import defaultdict, namedtuple


Node = namedtuple('Node', ['node_id', 'parent_id', 'depth'])
Edge = namedtuple('Edge', ['node_to', 'substring_start', 'substring_length'])


class SuffixTree:
    def __init__(self, text, suffix_array, lcp_array):
        self._text = text
        self._size = len(text)
        self._node_id_generator = it.count()
        self._adjacency_list = self._build_suffix_tree(suffix_array, lcp_array)

    def edges(self):
        for edge_labels in self._adjacency_list.values():
            for _, start_ix, length in edge_labels:
                yield self._text[start_ix:start_ix+length]

    def _build_suffix_tree(self, suffix_array, lcp_array):
        adjacency_list = defaultdict(list)
        current_node_id = next(self._node_id_generator)
        current_node = Node(node_id=current_node_id, parent_id=None, depth=0)
        nodes = {current_node_id: current_node}

        for suffix_start, lcp in zip(suffix_array, lcp_array):
            current_node = find_starting_node_for_next_edge(nodes, current_node, lcp)
            substring_start = suffix_start + current_node.depth
            if current_node.depth == lcp:
                current_node = self._add_new_leaf(adjacency_list, current_node, nodes, substring_start)
            else:
                edge = self._identify_edge(adjacency_list, current_node, substring_start)
                offset = lcp - current_node.depth
                middle_node = self._split_edge(current_node, edge, offset, nodes, adjacency_list)
                current_node = self._add_new_leaf(adjacency_list, middle_node, nodes, substring_start+offset)

        return dict(adjacency_list)

    def _split_edge(self, node, edge, offset, nodes, adjacency_list):
        end_node = nodes[edge.node_to]
        middle_node_id = next(self._node_id_generator)
        middle_node = Node(node_id=middle_node_id, parent_id=node.node_id, depth=node.depth+offset)
        nodes[middle_node_id] = middle_node
        nodes[end_node.node_id] = Node(node_id=end_node.node_id, parent_id=middle_node_id, depth=end_node.depth)

        first_edge = Edge(middle_node_id, edge.substring_start, offset)
        second_edge = Edge(edge.node_to, edge.substring_start+offset, edge.substring_length-offset)
        adjacency_list[middle_node_id].append(second_edge)
        adjacency_list[node.node_id].append(first_edge)
        adjacency_list[node.node_id].remove(edge)
        return middle_node

    def _add_new_leaf(self, adjacency_list, node, nodes, substring_start):
        leaf_id = next(self._node_id_generator)
        label_length = self._size - substring_start
        leaf = Node(node_id=leaf_id, parent_id=node.node_id, depth=node.depth+label_length)
        nodes[leaf_id] = leaf
        adjacency_list[node.node_id].append(Edge(leaf_id, substring_start, label_length))
        return leaf

    def _identify_edge(self, adjacency_list, current_node, substring_start):
        char = self._text[substring_start]
        for edge in adjacency_list[current_node.node_id]:
            if self._text[edge.substring_start] == char:
                return edge
        return None


def find_starting_node_for_next_edge(nodes, node, lcp):
    while node.depth > lcp:
        node = nodes[node.parent_id]
    return node


def convert_to_intlist(line):
    return [int(elem) for elem in line.split(', ')]


def main():
    reader = sys.stdin
    text = next(reader).rstrip()
    suffix_array = convert_to_intlist(next(reader))
    lcp_array = convert_to_intlist(next(reader))
    suffix_tree = SuffixTree(text, suffix_array, lcp_array)
    for edge_label in suffix_tree.edges():
        print(edge_label)


if __name__ == '__main__':
    main()
